File: level2/ex46.js
// 균형잡힌 괄호 문자열 확인 함수
function isBalanced(arr) {
  const arr1 = arr.filter(value => value === '(');
  const arr2 = arr.filter(value => value === ')');
  
  return arr1.length === arr2.length;
}

// 올바른 괄호 문자열 확인 함수
function isCorrect(arr) {
  const stack = [];
  
  for (const value of arr) {
      if (value === '(') {
          stack.push(value);
      } else {
          stack.pop();
      }
  }
  
  return stack.length === 0;
}

// 주어진 문자열을 두 균형잡힌 괄호 문자열로 분리하는 함수
function split(arr) {
  for (let i = 0; i < arr.length; i++) {
      const u = arr.slice(0, i + 1);
      const v = arr.slice(i + 1);
      
      if (isBalanced(u)) {
          return [u, v]
      }
  }
}

// 주어진 문자열의 괄호 방향을 뒤집어 반환하는 함수
function reverse(arr) {
  for (let i = 0; i < arr.length; i++) {
      if (arr[i] === '(') {
          arr[i] = ')';
      } else {
          arr[i] = '(';
      }
  }
  
  return arr;
}

// 균형잡힌 괄호 문자열 -> 올바른 괄호 문자열 변환 함수
function convert(w) {
  if (w.length === 0) {
      return w;
  }
  
  const [u, v] = split(w);
  
  if (isCorrect(u)) {
      return [...u, ...convert(v)].join('');
  } else {
      return ['(', ...convert(v), ...')', ...reverse(u).slice(1, -1)].join('');
  }
}

function solution(p) {
  return convert(p.split(''));
}

File: level2/ex46.py
# 균형잡힌 괄호 문자열 확인 함수
def is_balanced(w):
    return w.count('(') == w.count(')')

# 올바른 괄호 문자열 확인 함수
def is_correct(w):
    count = 0
    
    for value in w:
        if value == '(':
            count += 1
        else:
            count -= 1
        
        if count < 0:
            return False
    
    return count == 0
    
# 주어진 문자열을 두 균형잡힌 괄호 문자열로 분리하는 함수
def split(w):
    for i in range(len(w)):
        u, v = w[:i+1], w[i+1:]  
        if is_balanced(u):
            return u, v
    
# 주어진 문자열의 괄호 방향을 뒤집어 반환하는 함수
def reverse(w):
    reversed = ''
    for value in w:
        if value == '(':
            reversed += ')'
        else:
            reversed += '('
    return reversed
    
# 균형잡힌 괄호 문자열 -> 올바른 괄호 문자열 변환 함수
def convert(w):
    if w == '':
        return w

    u, v = split(w)

    if is_correct(u):
        return u + convert(v)
    else:
        return '(' + convert(v) + ')' + reverse(u[1:-1])

def solution(p):
    return convert(p)